drivers/crypto/crypto_it8xxx2_sha_v2.c

```diff
@@ -0,0 +1,325 @@
+					IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+			it8xxx2_sha256_module_calculation();
+		} else {
+			/* 0 <= Data length < 1KB */
+			while (rem_len) {
+				rem_len--;
+				chip_ctx.w_input[chip_ctx.w_input_index++] =
+								pkt->in_buf[in_buf_idx++];
+
+				/*
+				 * If fill full 64byte then execute HW calculation.
+				 * If not, will execute in later finish block.
+				 */
+				if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+					/* HW automatically load 64Bytes data from DLM */
+					sys_write8(IT8XXX2_SHAEXEC_64Byte,
+							IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+					it8xxx2_sha256_module_calculation();
+				}
+			}
+		}
+	}
+
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		/*
+		 * Handles the boundary case of rest data:
+		 * Because the last eight bytes are bit length field of sha256 rule.
+		 * If the data index >= 56, it needs to trigger HW to calculate,
+		 * then fill 0 data and the last eight bytes bit length, and calculate again.
+		 */
+		if (chip_ctx.w_input_index >= 56) {
+			/* HW automatically load 64Bytes data from DLM */
+			sys_write8(IT8XXX2_SHAEXEC_64Byte,
+					IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+			it8xxx2_sha256_module_calculation();
+			memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+				0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		}
+
+		/*
+		 * Since input data (big-endian) are copied 1byte by 1byte to
+		 * it8xxx2 memory (little-endian), so the bit length needs to
+		 * be transformed into big-endian format and then write to memory.
+		 */
+		chip_ctx.w_sha[15] = sys_cpu_to_be32(chip_ctx.total_len * 8);
+
+		/* HW automatically load 64Bytes data from DLM */
+		sys_write8(IT8XXX2_SHAEXEC_64Byte, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHAECR);
+		it8xxx2_sha256_module_calculation();
+
+		/* HW write back the hash result to DLM */
+		/* Set DLM address for input data */
+		sys_write8(((uint32_t)&chip_ctx.h) & 0xc0,
+				IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA0R);
+		sys_write8(((uint32_t)&chip_ctx.h) >> 8,
+				IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHADBA1R);
+
+		key = irq_lock();
+		/* Crypto use SRAM */
+		gctrl_regs->GCTRL_PMER3 |= IT8XXX2_GCTRL_SRAM_CRYPTO_USED;
+		sha_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+		sys_write8(sha_ctrl | IT8XXX2_SHAWB, IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHACR);
+
+		/* HW write back the hash result to DLM ~= 1us */
+		for (count = 0; count <= (SHA_SHA256_WRITE_BACK_TIMEOUT_US /
+		     SHA_SHA256_WAIT_NEXT_CLOCK_TIME_US); count++) {
+			/* Delay 15us */
+			gctrl_regs->GCTRL_WNCKR = IT8XXX2_GCTRL_WN65K;
+
+			if ((sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_SHASR)
+			      & IT8XXX2_SHAIS)) {
+				timeout = 0;
+				break;
+			}
+		}
+
+		if (timeout) {
```

Comment:
Is it a fatal error ? If so, shouldn't you just cleanup resources (e.g unlock irq) and return an error ?

Reply:
Just a reminder, I added the timeout message that was for solving infinite while(); loop problem. 

We expect that HW always can raise calculation done or write back done status. If HW is something wrong, status doesn't raise (although the chance is almost 0 and hasn't happened), then use k_oops() to reset our EC chip. We have the print message clue and can debug further to find the root cause.